File: scripts/linters/no_std_collections_maps.rb
def no_std_collections_maps(root_dir, excludes = [])
  pattern = root_dir.join('crates', '**', '*.rs').to_s
  errors = Dir.glob(pattern).sort.flat_map do |path|
    relative = Pathname.new(path).relative_path_from(root_dir).to_s
    next [] if excludes.include?(relative)

    find_std_map_usages(path, relative)
  end

  return true if errors.empty?

  puts 'Found uses of `std::collections::{HashMap, HashSet, BTreeMap, BTreeSet}`.'
  puts 'Use `indexmap::IndexMap` / `indexmap::IndexSet` instead:'
  errors.each do |err|
    puts "  #{err}"
  end
  false
end

BANNED_MAP_NAMES = %w[HashMap HashSet BTreeMap BTreeSet].freeze

def find_std_map_usages(path, relative)
  errors = []

  File.readlines(path).each_with_index do |raw, idx|
    code = raw.split('//', 2).first || raw

    code.scan(/\bstd::collections::(HashMap|HashSet|BTreeMap|BTreeSet)\b/) do |m|
      errors << "#{relative}:#{idx + 1}: use of `std::collections::#{m[0]}` (use `indexmap::#{indexmap_replacement(m[0])}` instead)"
    end

    code.scan(/\bstd::collections::\{([^}]*)\}/) do |m|
      m[0].split(',').each do |entry|
        name = entry.strip.split(/\s+as\s+/).first
        next unless BANNED_MAP_NAMES.include?(name)

        errors << "#{relative}:#{idx + 1}: import of `std::collections::#{name}` (use `indexmap::#{indexmap_replacement(name)}` instead)"
      end
    end
  end

  errors.uniq
end

def indexmap_replacement(name)
  case name
  when 'HashMap', 'BTreeMap' then 'IndexMap'
  when 'HashSet', 'BTreeSet' then 'IndexSet'
  end
end

File: scripts/linters/src/Linters/NoStdCollectionsMaps.php
<?php

declare(strict_types=1);

namespace Shirabe\Lint\Linters;

use Shirabe\Lint\Linter;
use Shirabe\Lint\Support\FileFinder;
use Shirabe\Lint\Support\Paths;

final class NoStdCollectionsMaps implements Linter
{
    private const BANNED_MAP_NAMES = ['HashMap', 'HashSet', 'BTreeMap', 'BTreeSet'];

    public function name(): string
    {
        return 'no_std_collections_maps';
    }

    public function failureIntro(): string
    {
        return "Found uses of `std::collections::{HashMap, HashSet, BTreeMap, BTreeSet}`.\n"
            . 'Use `indexmap::IndexMap` / `indexmap::IndexSet` instead:';
    }

    public function check(string $rootDir, array $excludes): array
    {
        $errors = [];

        foreach (FileFinder::rustFiles($rootDir) as $path) {
            $relative = Paths::relativeTo($rootDir, $path);
            if (in_array($relative, $excludes, true)) {
                continue;
            }

            array_push($errors, ...$this->findStdMapUsages($path, $relative));
        }

        return $errors;
    }

    /** @return list<string> */
    private function findStdMapUsages(string $path, string $relative): array
    {
        $errors = [];

        foreach (file($path) as $idx => $raw) {
            $code = explode('//', $raw, 2)[0];

            if (preg_match_all('/\bstd::collections::(HashMap|HashSet|BTreeMap|BTreeSet)\b/', $code, $m)) {
                foreach ($m[1] as $name) {
                    $errors[] = "{$relative}:" . ($idx + 1) . ": use of `std::collections::{$name}` (use `indexmap::" . self::indexmapReplacement($name) . '` instead)';
                }
            }

            if (preg_match_all('/\bstd::collections::\{([^}]*)\}/', $code, $m)) {
                foreach ($m[1] as $group) {
                    foreach (explode(',', $group) as $entry) {
                        $name = preg_split('/\s+as\s+/', trim($entry))[0];
                        if (!in_array($name, self::BANNED_MAP_NAMES, true)) {
                            continue;
                        }
                        $errors[] = "{$relative}:" . ($idx + 1) . ": import of `std::collections::{$name}` (use `indexmap::" . self::indexmapReplacement($name) . '` instead)';
                    }
                }
            }
        }

        return array_values(array_unique($errors));
    }

    private static function indexmapReplacement(string $name): string
    {
        return match ($name) {
            'HashMap', 'BTreeMap' => 'IndexMap',
            'HashSet', 'BTreeSet' => 'IndexSet',
            default => throw new \LogicException("unexpected map name: {$name}"),
        };
    }
}
